File: ji/proj_2/app/src/main/java/com/example/proj_2/tab2/OpenGalleryFolderActivity.java
package com.example.proj_2.tab2;

import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.proj_2.R;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

public class OpenGalleryFolderActivity extends AppCompatActivity {
    Intent intent = new Intent();
    Long bucketId = intent.getLongExtra("bucketId", -100);
    String bucketName = intent.getStringExtra("bucketName");
    public ArrayList<MediaFileData> list = new ArrayList<MediaFileData>();

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_opengalleryfolder);


        List<MediaFileData> ImageDataset = getFileList(this).stream().filter(a -> a.bucketId == bucketId).collect(Collectors.toList());


        //it.bucketId == bucketId;
        // }
        RecyclerView recyclerView1 = findViewById(R.id.opened_gallery);
        recyclerView1.setLayoutManager(new GridLayoutManager(this.getBaseContext(), 4));
        recyclerView1.setAdapter(new InGalleryImageAdapter(list));
// return MediaFileData which contains id, dateTaken, displayName, uri
        // Use MediaStore API

    }

    ArrayList<MediaFileData> getFileList(Context context){
        ArrayList<MediaFileData> folderList = new ArrayList<MediaFileData>();
        String[] projection = {MediaStore.Files.FileColumns._ID, MediaStore.Files.FileColumns.DISPLAY_NAME,
                MediaStore.Files.FileColumns.DATE_TAKEN,
                MediaStore.Files.FileColumns.BUCKET_DISPLAY_NAME,
                MediaStore.Files.FileColumns.BUCKET_ID
        };


        String sortOrder = MediaStore.Files.FileColumns.DATE_TAKEN + " " + "DESC";

        //ContentResolver contentResolver1 = context.getContentResolver();
        ContentResolver contentResolver1 = context.getContentResolver();
        Cursor cursor = contentResolver1.query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection, null, null, sortOrder);
        //Cursor cursor = getApplicationContext().getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection, null, null, sortOrder);

        // cursor.use {
        int idColumn = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns._ID);
        int dateTakenColumn = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.DATE_TAKEN);
        int displayNameColumn = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.DISPLAY_NAME);
        int bucketIDColumn = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.BUCKET_ID);
        int bucketNameColumn = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.BUCKET_DISPLAY_NAME);
        while (cursor.moveToNext()) {
            Long id = cursor.getLong(idColumn);
            Date dateTaken = new Date(cursor.getLong(dateTakenColumn));
            String displayName = cursor.getString(displayNameColumn);
            Uri contentUri = Uri.withAppendedPath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, id.toString());
            Long bucketID = cursor.getLong(bucketIDColumn);
            String bucketName = cursor.getString(bucketNameColumn);

            MediaFileData MDF = new MediaFileData(bucketID, id, dateTaken, displayName, contentUri, bucketName);
            folderList.add(MDF);

        }
        // }

        return folderList;
    }


}

File: ji/project1_final/app/src/main/java/com/example/project1_final/OpenGalleryFolderActivity.kt
package com.example.project1_final

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.project1_final.adapter.InGalleryImageAdapter
import com.example.project1_final.fragments.GalleryFragmentTab
import com.example.project1_final.model.MediaFileData
import kotlinx.android.synthetic.main.item_folder_image.*
import kotlinx.android.synthetic.main.activity_opengalleryfolder.*
import java.util.*

class OpenGalleryFolderActivity:AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?){
        super.onCreate(savedInstanceState)

        val bucketId=intent.getLongExtra("bucketId", -100)
        val bucketName=intent.getStringExtra("bucketName")

        actionBar?.title = bucketName;
        supportActionBar?.title = bucketName;
        setContentView(R.layout.activity_opengalleryfolder)

        val ImageDataset = getFileList(this, GalleryFragmentTab.MediaStoreFileType.IMAGE).filter{ it.bucketId == bucketId }

        val recyclerView: RecyclerView = opened_gallery

        recyclerView.layoutManager = GridLayoutManager(this, 4)
        recyclerView.adapter = InGalleryImageAdapter(this, ImageDataset)
    }

    fun getFileList(context: Context, type: GalleryFragmentTab.MediaStoreFileType): List<MediaFileData> {

        val folderList = mutableListOf<MediaFileData>()
        val projection = arrayOf(
                MediaStore.Files.FileColumns._ID,
                MediaStore.Files.FileColumns.DISPLAY_NAME,
                MediaStore.Files.FileColumns.DATE_TAKEN,
                MediaStore.Files.FileColumns.BUCKET_DISPLAY_NAME,
                MediaStore.Files.FileColumns.BUCKET_ID
        )

        val sortOrder = "${MediaStore.Files.FileColumns.DATE_TAKEN} DESC"

        val cursor = context.contentResolver.query(
                type.externalContentUri,
                projection,
                null,
                null,
                sortOrder
        )

        cursor?.use {
            val idColumn = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns._ID)
            val dateTakenColumn =
                    cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.DATE_TAKEN)
            val displayNameColumn =
                    cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.DISPLAY_NAME)
            val bucketIDColumn =
                    cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.BUCKET_ID)
            val bucketNameColumn =
                    cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.BUCKET_DISPLAY_NAME)
            while (cursor.moveToNext()) {
                val id = cursor.getLong(idColumn)
                val dateTaken = Date(cursor.getLong(dateTakenColumn))
                val displayName = cursor.getString(displayNameColumn)
                val contentUri = Uri.withAppendedPath(
                        type.externalContentUri,
                        id.toString()
                )
                val bucketID = cursor.getLong(bucketIDColumn)
                val bucketName = cursor.getString(bucketNameColumn)

                Log.d(
                        "test",
                        "id: $id, display_name: $displayName, date_taken: $dateTaken, content_uri: $contentUri\n"
                )

                val MDF = MediaFileData(id, dateTaken, displayName, contentUri, bucketID, bucketName)
                folderList.add(MDF)

            }
        }

        return folderList
    }

}